rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/Rename.rsc

```diff
@@ -0,0 +1,400 @@
+                // Collection field uses; any location which is of set or list type, where the label of the collection element equals the name under the cursor
+              , <findSmallestContaining({l | l <- ws.facts, at := ws.facts[l], (at is aset || at is alist) && at.elmType.alabel? && at.elmType.alabel == cursorName}, cursorLoc), collectionField()>
+                // Module name declaration, where the cursor location is in the module header
+              , <flatMap(locationOfName(parseModuleWithSpacesCached(cursorLoc.top).top.header), Maybe[loc](loc nameLoc) { return isContainedIn(cursorLoc, nameLoc) ? just(nameLoc) : nothing(); }), moduleName()>
+            }
+    };
+
+    if (locsContainingCursor == {}) {
+        throw unsupportedRename("Cannot find type information in TPL for <cursorLoc>");
+    }
+
+    loc c = min(locsContainingCursor.l);
+    Cursor cur = cursor(use(), |unknown:///|, "");
+    switch (locsContainingCursor[c]) {
+        case {moduleName(), *_}: {
+            cur = cursor(moduleName(), c, cursorName);
+        }
+        case {def(), *_}: {
+            // Cursor is at a definition
+            cur = cursor(def(), c, cursorName);
+        }
+        case {use(), *_}: {
+            if (d <- ws.useDef[c], just(amodule(_)) := getFact(ws, d)) {
+                // Cursor is at an import
+                cur = cursor(moduleName(), c, cursorName);
+            } else if (u <- ws.useDef<0>, u.begin <= cursorLoc.begin && u.end > cursorLoc.end) {
+                // Cursor is at a qualified name
+                cur = cursor(moduleName(), c, cursorName);
+            } else if (size(getDefs(ws, c) & ws.defines.defined) > 0) {
+                // The cursor is at a use with corresponding definitions.
+                cur = cursor(use(), c, cursorName);
+            } else if (just(at) := getFact(ws, c)) {
+                if (aparameter(cursorName, _) := at) {
+                    // The cursor is at a type parameter
+                    cur = cursor(typeParam(), c, cursorName);
+                } else if (at.alabel == cursorName) {
+                    // The cursor is at a collection field
+                    cur = cursor(collectionField(), c, cursorName);
+                }
+            }
+        }
+        case {k}: {
+            cur = cursor(k, c, cursorName);
+        }
+        default:
+            throw unsupportedRename("Unsupported cursor type <locsContainingCursor[c]>");
+    }
+
+    if (cur.l.scheme == "unknown") throw unexpectedFailure("Could not find cursor location.");
+
+    return <cur, ws>;
+}
+
+private bool containsName(loc l, str name) {
+    // If we do not find any occurrences of the name under the cursor in a module,
+    // we are not interested in it at all, and will skip loading its TPL.
+    cursorAsName = [Name] name;
+    escapedCursorAsName = startsWith(name, "\\") ? name : [Name] "\\<name>";
```

Comment:
Did you intend the following?
```suggestion
    escapedCursorAsName = startsWith(name, "\\") ? cursorAsName : [Name] "\\<name>";
```

Reply:
Indeed. I'll add some types here as well, so the typechecker could have caught this.